File: leetcode_python/Binary_Search_Tree/build-binary-expression-tree-from-infix-expression.py
"""

1597. Build Binary Expression Tree From Infix Expression
Hard

A binary expression tree is a kind of binary tree used to represent arithmetic expressions. Each node of a binary expression tree has either zero or two children. Leaf nodes (nodes with 0 children) correspond to operands (numbers), and internal nodes (nodes with 2 children) correspond to the operators '+' (addition), '-' (subtraction), '*' (multiplication), and '/' (division).

For each internal node with operator o, the infix expression that it represents is (A o B), where A is the expression the left subtree represents and B is the expression the right subtree represents.

You are given a string s, an infix expression containing operands, the operators described above, and parentheses '(' and ')'.

Return any valid binary expression tree, which its in-order traversal reproduces s after omitting the parenthesis from it (see examples below).

Please note that order of operations applies in s. That is, expressions in parentheses are evaluated first, and multiplication and division happen before addition and subtraction.

Operands must also appear in the same order in both s and the in-order traversal of the tree.

 

Example 1:


Input: s = "3*4-2*5"
Output: [-,*,*,3,4,2,5]
Explanation: The tree above is the only valid tree whose inorder traversal produces s.
Example 2:


Input: s = "2-3/(5*2)+1"
Output: [+,-,1,2,/,null,null,null,null,3,*,null,null,5,2]
Explanation: The inorder traversal of the tree above is 2-3/5*2+1 which is the same as s without the parenthesis. The tree also produces the correct result and its operands are in the same order as they appear in s.
The tree below is also a valid binary expression tree with the same inorder traversal as s, but it not a valid answer because it does not evaluate to the same value.

The third tree below is also not valid. Although it produces the same result and is equivalent to the above trees, its inorder traversal does not produce s and its operands are not in the same order as s.

Example 3:

Input: s = "1+2+3+4+5"
Output: [+,+,5,+,4,null,null,+,3,null,null,1,2]
Explanation: The tree [+,+,5,+,+,null,null,1,2,3,4] is also one of many other valid trees.
 

Constraints:

1 <= s.length <= 1000
s consists of digits and the characters '+', '-', '*', and '/'.
Operands in s are exactly 1 digit.
It is guaranteed that s is a valid expression.

"""

# V0
# IDEA : LC 224 Basic Calculator
class Solution(object):

    def help(self, numSt, opSt):
        right = numSt.pop()
        left = numSt.pop()
        # Node(val=op, left=lhs, right=rhs)
        return Node(opSt.pop(), left, right)

    def expTree(self, s):
        # hashmap for operator ordering
        pr = {'*': 1, '/': 1, '+': 2, '-': 2, ')': 3, '(': 4}
        numSt = []
        opSt = []
        i = 0
        while i < len(s):
            c = s[i]
            i += 1
            # check if int(c) if string
            if c.isnumeric():
                numSt.append(Node(c))
            else:                
                if c == '(':
                    opSt.append('(')
                else:
                    while(len(opSt) > 0 and pr[c] >= pr[opSt[-1]]):
                        numSt.append(self.help(numSt, opSt))
                    if c == ')':
                        opSt.pop() # Now what remains is the closing bracket ')'
                    else:
                        opSt.append(c)
        while len(opSt) > 0:
            numSt.append(self.help(numSt, opSt))
        print (">>> numSt = {}, opSt = {}".format(str(numSt), opSt))
        return numSt.pop()

# V0'
# IDEA : RECURSIVE
class Solution:
    def expTree(self, s):
        n = len(s)
        if n == 1:
            return Node(s)

        fstOpIdx = None
        kets = 0
        for i in range(n-1, 0, -1):
            if s[i] == ")":
                kets += 1
            elif s[i] == "(":
                kets -= 1
            elif kets == 0:
                if s[i] in "+-":
                    fstOpIdx = i
                    break
                elif s[i] in "*/" and fstOpIdx is None:
                    fstOpIdx = i
        if fstOpIdx is None:
            return self.expTree(s[1:-1])
        rtNd = Node(s[fstOpIdx])
        rtNd.left = self.expTree(s[:fstOpIdx])
        rtNd.right = self.expTree(s[fstOpIdx+1:])
        return rtNd

# V1
# https://leetcode.com/problems/build-binary-expression-tree-from-infix-expression/discuss/864596/Python-Standard-parser-implementation
class Solution:
    # Standard parser implementation based on this BNF
    #   s := expression
    #   expression := term | term { [+,-] term] }
    #   term := factor | factor { [*,/] factor] }
    #   factor :== digit | '(' expression ')'
    #   digit := [0..9]
    
    def expTree(self, s):
        tokens = collections.deque(list(s))
        return self.parse_expression(tokens)

    def parse_expression(self, tokens):
        lhs = self.parse_term(tokens)
        while len(tokens) > 0 and tokens[0] in ['+', '-']:
            op = tokens.popleft()
            rhs = self.parse_term(tokens)
            lhs = Node(val=op, left=lhs, right=rhs)
        return lhs
    
    def parse_term(self, tokens):
        lhs = self.parse_factor(tokens)
        while len(tokens) > 0 and tokens[0] in ['*', '/']:
            op = tokens.popleft()
            rhs = self.parse_factor(tokens)
            lhs = Node(val=op, left=lhs, right=rhs)
        return lhs

    def parse_factor(self, tokens):
        if tokens[0] == '(':
            tokens.popleft() # consume '('
            node = self.parse_expression(tokens)
            tokens.popleft() # consume ')'
            return node
        else:
            # Single operand
            token = tokens.popleft()
            return Node(val=token)

# V1'
# IDEA : LC 224 Basic Calculator
# https://leetcode.com/problems/build-binary-expression-tree-from-infix-expression/discuss/944215/Python-Solution-Similar-to-Basic-Calculator
class Solution(object):

    def apply(self, numSt, opSt):
        right = numSt.pop()
        left = numSt.pop()
        return Node(opSt.pop(), left, right)

    def expTree(self, s):
        pr = {'*': 1, '/': 1, '+': 2, '-': 2, ')': 3, '(': 4}
        numSt, opSt = [], []
        i = 0
        while (i < len(s)):
            c = s[i]
            i += 1
            if c.isnumeric():
                numSt.append(Node(c))
            else:                
                if c in '(':
                    opSt.append('(')
                else:
                    while(len(opSt) > 0 and pr[c] >= pr[opSt[-1]]):
                        numSt.append(self.apply(numSt, opSt))
                    if (c == ')'):
                        opSt.pop() # Now what remains is the closing bracket ')'
                    else:
                        opSt.append(c)
        while len(opSt) > 0:
            numSt.append(self.apply(numSt, opSt))
        return numSt.pop()

# V1''
# IDEA : RECURSIVE
# https://leetcode.com/problems/build-binary-expression-tree-from-infix-expression/discuss/985360/Python-3-recursion-and-easy-to-understand
class Solution:
    def expTree(self, s):
        n = len(s)
        if n == 1:
            return Node(s)

        fstOpIdx = None
        kets = 0
        for i in range(n-1, 0, -1):
            if s[i] == ")":
                kets += 1
            elif s[i] == "(":
                kets -= 1
            elif kets == 0:
                if s[i] in "+-":
                    fstOpIdx = i
                    break
                elif s[i] in "*/" and fstOpIdx is None:
                    fstOpIdx = i
        if fstOpIdx is None:
            return self.expTree(s[1:-1])
        rtNd = Node(s[fstOpIdx])
        rtNd.left = self.expTree(s[:fstOpIdx])
        rtNd.right = self.expTree(s[fstOpIdx+1:])
        return rtNd

# V1'''
# https://leetcode.com/problems/build-binary-expression-tree-from-infix-expression/discuss/862421/Python-O(n)-single-pass-Handles-multiple-digits-and-negative-numbers
class Solution:
    def expTree(self, expression: str) -> 'Node':
        precedence = {'/':2, '*':2, '+':1, '-':1}
        ops = {'/','*','+','-'}
        operator = []
        operand = [Node('0')]
        
        i = 0
        while i<len(expression):
            ch = expression[i]

            if ch.isdigit():
                start = i
                while i<len(expression) and expression[i].isdigit():
                    i += 1
                num = expression[start:i]
                operand.append(Node(num))
                i -= 1
                
            elif ch in ops:
                while len(operator) and operator[-1].val!='(' and precedence[operator[-1].val] >= precedence[ch]:
                    self.process(operator, operand)
                operator.append(Node(ch))

            elif ch == '(': 
                operator.append(Node(ch)); 
                if expression[i+1] == '-':
                    operand.append(Node('0'))

            elif ch == ')':
                while operator[-1].val != '(':
                    self.process(operator, operand)
                operator.pop()
            
            i += 1

        while len(operator):
            self.process(operator,operand)
        
        return operand.pop()
    
    def process(self,operator_stack, operand_stack):
        node2 = operand_stack.pop()
        node1 = operand_stack.pop()
        node = operator_stack.pop()

        node.left = node1
        node.right = node2

        operand_stack.append(node)

# V2

File: leetcode_java/src/main/java/LeetCodeJava/BinarySearchTree/BuildBinaryExpressionTreeFromInfixExpression.java
package LeetCodeJava.BinarySearchTree;

// https://leetcode.com/problems/build-binary-expression-tree-from-infix-expression/

import java.util.ArrayDeque;
import java.util.Deque;

/**
 *  1597. Build Binary Expression Tree From Infix Expression
 *  Hard
 *
 *  A binary expression tree is a kind of binary tree used to represent arithmetic
 *  expressions. Each node has either zero or two children. Leaf nodes correspond
 *  to operands (numbers), and internal nodes correspond to the operators
 *  '+', '-', '*' and '/'.
 *
 *  For each internal node with operator o, the infix expression it represents is
 *  (A o B), where A is the expression of the left subtree and B that of the right.
 *
 *  You are given a string s, an infix expression containing operands, the
 *  operators described above, and parentheses '(' and ')'.
 *
 *  Return any valid binary expression tree whose in-order traversal reproduces s
 *  after omitting the parentheses. Order of operations applies: parentheses
 *  first, then '*' and '/', then '+' and '-'. Operands must appear in the same
 *  order in s and in the in-order traversal.
 *
 *  Example 1:
 *   Input: s = "3*4-2*5"
 *   Output: [-,*,*,3,4,2,5]
 *
 *  Example 2:
 *   Input: s = "2-3/(5*2)+1"
 *   Output: [+,-,1,2,/,null,null,null,null,3,*,null,null,5,2]
 *
 *  Constraints:
 *   1 <= s.length <= 1000
 *   s consists of digits and the characters '+', '-', '*', '/', '(' and ')'.
 *   Operands in s are exactly 1 digit.
 *   It is guaranteed that s is a valid expression.
 */
public class BuildBinaryExpressionTreeFromInfixExpression {

    // LC "Node" for this problem (char val + left/right), defined locally since
    // the shared LeetCodeJava.DataStructure.Node is the clone-graph shape.
    public static class NodeX {
        public char val;
        public NodeX left;
        public NodeX right;

        public NodeX() {
            this.val = ' ';
        }

        public NodeX(char val) {
            this.val = val;
        }

        public NodeX(char val, NodeX left, NodeX right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    // V0
    // IDEA: shunting-yard (same skeleton as LC 224 Basic Calculator) with two
    //       stacks: operands become leaves, and whenever an operator of >= 
    //       precedence is already on top we pop and build a subtree.
    /**
     * time = O(n)
     * space = O(n)
     */
    public NodeX expTree(String s) {
        Deque<NodeX> nums = new ArrayDeque<>();
        Deque<Character> ops = new ArrayDeque<>();

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);

            if (Character.isDigit(c)) {
                nums.push(new NodeX(c));
            } else if (c == '(') {
                ops.push(c);
            } else if (c == ')') {
                while (!ops.isEmpty() && ops.peek() != '(') {
                    build(nums, ops);
                }
                ops.pop(); // drop '('
            } else {
                // operator: flush everything with >= precedence
                while (!ops.isEmpty() && ops.peek() != '('
                        && precedence(ops.peek()) >= precedence(c)) {
                    build(nums, ops);
                }
                ops.push(c);
            }
        }

        while (!ops.isEmpty()) {
            build(nums, ops);
        }
        return nums.isEmpty() ? null : nums.pop();
    }

    private void build(Deque<NodeX> nums, Deque<Character> ops) {
        NodeX right = nums.pop();
        NodeX left = nums.pop();
        nums.push(new NodeX(ops.pop(), left, right));
    }

    private int precedence(char op) {
        if (op == '*' || op == '/') {
            return 2;
        }
        if (op == '+' || op == '-') {
            return 1;
        }
        return 0;
    }

    // cursor used by the V1 recursive-descent parser
    private int pos_1;

    // V1
    // IDEA: recursive-descent parser over the grammar
    //         expr   := term (('+'|'-') term)*
    //         term   := factor (('*'|'/') factor)*
    //         factor := digit | '(' expr ')'
    //       Precedence is encoded in the grammar itself instead of in a stack,
    //       and the left-folding loop gives left associativity.
    /**
     * time = O(n)
     * space = O(n)   // recursion depth on nested parentheses
     */
    public NodeX expTree_1(String s) {
        pos_1 = 0;
        return parseExpr_1(s);
    }

    private NodeX parseExpr_1(String s) {
        NodeX left = parseTerm_1(s);
        while (pos_1 < s.length() && (s.charAt(pos_1) == '+' || s.charAt(pos_1) == '-')) {
            char op = s.charAt(pos_1++);
            NodeX right = parseTerm_1(s);
            left = new NodeX(op, left, right);
        }
        return left;
    }

    private NodeX parseTerm_1(String s) {
        NodeX left = parseFactor_1(s);
        while (pos_1 < s.length() && (s.charAt(pos_1) == '*' || s.charAt(pos_1) == '/')) {
            char op = s.charAt(pos_1++);
            NodeX right = parseFactor_1(s);
            left = new NodeX(op, left, right);
        }
        return left;
    }

    private NodeX parseFactor_1(String s) {
        char c = s.charAt(pos_1);
        if (c == '(') {
            pos_1++;                      // consume '('
            NodeX inner = parseExpr_1(s);
            pos_1++;                      // consume ')'
            return inner;
        }
        pos_1++;
        return new NodeX(c);
    }

    // V2
    // IDEA: divide & conquer on the substring. Strip redundant wrapping parens,
    //       then split at the RIGHTMOST top-level (depth 0) operator of the
    //       LOWEST precedence -- that operator is evaluated last, so it is the
    //       root -- and recurse on the two halves.
    /**
     * time = O(n^2) worst case (rescans each substring)
     * space = O(n)
     */
    public NodeX expTree_2(String s) {
        if (s == null || s.isEmpty()) {
            return null;
        }
        return buildDC_2(s, 0, s.length() - 1);
    }

    private NodeX buildDC_2(String s, int lo, int hi) {
        // "(...)" wrapping the whole range adds nothing -> peel it off
        while (lo < hi && s.charAt(lo) == '(' && closesAt_2(s, lo, hi)) {
            lo++;
            hi--;
        }
        if (lo == hi) {
            return new NodeX(s.charAt(lo));
        }

        int addSub = -1;
        int mulDiv = -1;
        int depth = 0;
        for (int i = lo; i <= hi; i++) {
            char c = s.charAt(i);
            if (c == '(') {
                depth++;
            } else if (c == ')') {
                depth--;
            } else if (depth == 0) {
                if (c == '+' || c == '-') {
                    addSub = i;           // keep the rightmost -> left associative
                } else if (c == '*' || c == '/') {
                    mulDiv = i;
                }
            }
        }

        int split = addSub != -1 ? addSub : mulDiv;
        return new NodeX(s.charAt(split),
                buildDC_2(s, lo, split - 1),
                buildDC_2(s, split + 1, hi));
    }

    // does the '(' at index lo close exactly at index hi ?
    private boolean closesAt_2(String s, int lo, int hi) {
        int depth = 0;
        for (int i = lo; i <= hi; i++) {
            if (s.charAt(i) == '(') {
                depth++;
            } else if (s.charAt(i) == ')') {
                depth--;
                if (depth == 0) {
                    return i == hi;
                }
            }
        }
        return false;
    }
}
